lib/reverse_words.rb

```diff
@@ -1,28 +1,75 @@
-# A method to reverse each word in a sentence, in place.
-def reverse_words(my_words)
-  if my_words.nil?
-    return nil
-  else
-    i = 0
-    until my_words[i] == nil
-      if my_words[i] != " "
-        k = i
-        until my_words[k] == " " || my_words[k] == nil
-          k += 1
-        end
-        j = k-1
-        while i < j
-          memo = my_words[i]
-          my_words[i] = my_words[j]
-          my_words[j] = memo
-          i = i+1
-          j = j-1
-        end
-        i = k+1
-      else
-        i += 1
-      end
+# Time: O(n)
+# Space: O(1)
+def reverse_words(message)
+  return nil if !message
+  raise ArgumentError if message.class != String
+
+  i = 0
+  while i < (message.length - 1)
+
+    until message[i] != " " || i == (message.length - 1)
+      i += 1
     end
-    return my_words
+    j = i
+    until message[j + 1] == " " || j == (message.length - 1)
+      j += 1
+    end
+
+    message[i..j] = reverse_string_in_place(message[i..j])
```

Comment:
The input parameter to reverse_string_in_place takes up extra space since message[i..j] creates a new string object and then passes it in.

Reply:
OH! yes, thank you!